In [1]:
import pandas as pd

In [10]:
train = pd.read_csv("data/train.csv")
test = pd.read_csv("data/test.csv")

In [7]:
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

We have following 10 features of each passenger:

* *Pclass* is the Ticket-class: first(1), second(2), and third(3) class tickets. This is an ordinal integer feature.

* *Name* is the name of passenger. Although the name itself does not reveal too much information to prediction of survival, there is some titles attached to names that can indicate a certain group. For example, Mrs is an indication of marriage. This is a categorical text string feature.

* *Sex* is the gender of passenger. Either female or male. This is a categorical text string feature.

* *Age* is the integer age of passenger. There might be some NaN values in this column. This is an integer feature.

* *SibSp* is the number of siblings and spouse aboard the Titanic. Sibling includes brother, sister, stepbrother, and stepsister, while spouse includes husband or wife (mistresses and fiances were ignored). This is an ordinal integer feature.

* *Parch* is the number of parents and children aboard the Titanic. This is also an ordinal integer feature.

* *Ticket* is a character string that gives the ticket number.

* *Fare* is a float feature showing how much money the passenger paid for their trip.

* *Cabin* is the cabin number of each passenger. There are NaN in this column. This is also another string feature.

* *Embarked* is the port of embarkation as a categorical character feature.

In summary we have 1 floating point feature (fare), 1 integer variable (age), 3 ordinal integer features (plcass, sibsp, parch), 2 categorical features (sex, embarked), and 3 text string features (ticket, cabin, name).

In [8]:
train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

## Missing values

Take a look at missing values in training data:

In [9]:
print(train.isnull().sum())

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


It can be seen that 177 values were missing in Age, 687 values were missing in Cabin, and 2 values were missing in Embarked.

Let's fill in the empty values with the available information. Firstly, let's take a look at missing values of Embarked:

In [12]:
df = train[train["Embarked"].isnull()]
df

PassengerId  Survived  Pclass                                       Name  \
61            62         1       1                        Icard, Miss. Amelie   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket  Fare Cabin Embarked  
61   female  38.0      0      0  113572  80.0   B28      NaN  
829  female  62.0      0      0  113572  80.0   B28      NaN

Then we find similar passengers:

In [22]:
train[(train["Sex"]=="female") & (train["SibSp"]==0) & (train["Parch"]==0) & (train["Fare"]>70.0) & (train["Fare"]<90.0)]

PassengerId  Survived  Pclass  \
61            62         1       1   
218          219         1       1   
256          257         1       1   
257          258         1       1   
290          291         1       1   
310          311         1       1   
504          505         1       1   
627          628         1       1   
759          760         1       1   
829          830         1       1   

                                                  Name     Sex   Age  SibSp  \
61                                 Icard, Miss. Amelie  female  38.0      0   
218                              Bazzani, Miss. Albina  female  32.0      0   
256                     Thorne, Mrs. Gertrude Maybelle  female   NaN      0   
257                               Cherry, Miss. Gladys  female  30.0      0   
290                       Barber, Miss. Ellen "Nellie"  female  26.0      0   
310                     Hays, Miss. Margaret Bechstein  female  24.0      0   
504                              Maioni, Miss. Roberta  female  16.0      0   
627                      Longley, Miss. Gretchen Fiske  female  21.0      0   
759  Rothes, the Countess. of (Lucy Noel Martha Dye...  female  33.0      0   
829          Stone, Mrs. George Nelson (Martha Evelyn)  female  62.0      0   

     Parch    Ticket     Fare Cabin Embarked  
61       0    113572  80.0000   B28      NaN  
218      0     11813  76.2917   D15        C  
256      0  PC 17585  79.2000   NaN        C  
257      0    110152  86.5000   B77        S  
290      0     19877  78.8500   NaN        S  
310      0     11767  83.1583   C54        C  
504      0    110152  86.5000   B79        S  
627      0     13502  77.9583    D9        S  
759      0    110152  86.5000   B77        S  
829      0    113572  80.0000   B28      NaN

We can see that the similar passengers either embarked from C or S. Let's make it S here

In [38]:
train.ix[61,"Embarked"] = "S"
train.ix[829,"Embarked"] = "S"

Confirm there is no null values in column of Embarked now:

In [41]:
print(train.isnull().sum())

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64


In [42]:
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

There is a good explanation of [feature extraction and derivation](https://www.kaggle.com/headsortails/pytanic) in Kaggle. But to make things as simple as possible, let's just simply select the following features: Sex, Age, SibSp, ParCh, Fare, Embarked.

## Preparing the data

Since not all classifiers can handle string input, it's a better practice to adjust the column types to integers.